plugins/out_websocket/websocket.c

```diff
@@ -0,0 +1,283 @@
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <fluent-bit/flb_info.h>
+#include <fluent-bit/flb_output.h>
+#include <fluent-bit/flb_utils.h>
+#include <fluent-bit/flb_network.h>
+#include <fluent-bit/flb_time.h>
+#include <fluent-bit/flb_upstream.h>
+#include <fluent-bit/flb_sha512.h>
+#include <fluent-bit/flb_pack.h>
+#include <fluent-bit/flb_sds.h>
+#include <fluent-bit/flb_http_client.h>
+#include <msgpack.h>
+
+#include "websocket.h"
+#include "websocket_conf.h"
+struct flb_output_plugin out_websocket_plugin;
+
+#define SECURED_BY "Fluent Bit"
+
+
+static int flb_ws_handshake(struct flb_upstream_conn *u_conn,
+                                    struct flb_out_ws *ctx)
+{
+    int ret;
+    size_t bytes_sent;
+    struct flb_http_client *c;
+
+    if (!u_conn) {
+        flb_error("[output_ws] upstream connection error");
+        return -1;
+    }
+	
+    /* Compose HTTP Client request */
+    c = flb_http_client(u_conn, FLB_HTTP_GET, ctx->uri,
+                        NULL, 0, NULL, 0, NULL, 0);
+    if(!c) {
+       flb_upstream_conn_release(u_conn);
+       return -1;
+    }
+
+    flb_http_buffer_size(c, ctx->buffer_size);
+    flb_http_add_header(c, "Upgrade", 7, "websocket", 9);
+    flb_http_add_header(c, "Connection", 10, "Upgrade", 7);
+    flb_http_add_header(c, "Sec-WebSocket-Key", 17, "dGhlIHNhbXBsZSBub25jZQ==", 24);
+    flb_http_add_header(c, "Sec-WebSocket-Version", 21, "13", 2);
+	
+    /* Perform request*/
+    ret = flb_http_do(c, &bytes_sent);
+	
+    if (ret != 0 || c->resp.status != 101) {
+        if (c->resp.payload_size > 0) {
+            flb_debug("[output_ws] Websocket Server Response\n%s",
+                c->resp.payload);
+        }
+        flb_http_client_destroy(c);
+        flb_upstream_conn_release(u_conn);
+        flb_debug("[out_ws] Http Get Operation ret = %i, http resp = %i", ret, c->resp.status);
+        return -1;
+    }
+    flb_http_client_destroy(c);
+    return 0;
+}
+
+static void flb_ws_mask(char *data, int len, char *mask)
+{
+    int i;
+    for (i=0;i<len;++i) {
+        *(data+i) ^= *(mask+(i%4));
+    }
+}
+
+static int flb_ws_sendDataFrameHeader(struct flb_upstream_conn *u_conn, 
+                                      struct flb_out_ws *ctx, const void *data, size_t bytes)
+{		
+    int ret = -1;	
+    char* data_frame_head;
+    size_t bytes_sent;
+    int data_frame_head_len = 0;
+    //TODO use random function to generate masking_key
+    char masking_key[4] = {0x12, 0x34, 0x56, 0x78};
+    unsigned long long payloadSize = bytes;
+    
+    flb_ws_mask((char *)data, payloadSize, masking_key);
+    if(payloadSize < 126)
+    {
+        data_frame_head = (char *)flb_malloc(6);
```

Comment:
validate return of flb_malloc, if it fails do flb_errno() and return the exception